resources/sdk/purecloudjava/tests/ApiClientRetryTest.java

```diff
@@ -0,0 +1,445 @@
+        Assert.assertFalse(result, "Even though Status Code is 502, it will return false because MaxRetryTime is set to Zero by default");
+        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 0 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 100);
+        stopwatch.stop();
+    }
+
+    @Test
+    public void shouldRetryTestWith_503_And_RetryConfig() throws IOException {
+
+        ApiClientConnectorResponse response = getConnectorResponse(503, Collections.emptyMap());
+
+        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
+        retryConfiguration.setBackoffInterval(6000L);
+        retryConfiguration.setDefaultDelay(3000L);
+        retryConfiguration.setMaxRetryTime(10);
+        retryConfiguration.setMaxRetriesBeforeBackoff(0);
+
+        ApiClient.Retry retry = new ApiClient.Retry(retryConfiguration);
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        boolean result = retry.shouldRetry(response);
+        Assert.assertTrue(result, "Since Status Code is 503 and maxRetriesBeforeBackoff is Zero, backoff block will be executed and returns true");
+        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 3000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 3100);
+        stopwatch.stop();
+    }
+
+    @Test
+    public void shouldRetryTestWith_504_And_No_RetryAfter_Header() throws IOException {
+
+        ApiClientConnectorResponse response = getConnectorResponse(504, Collections.emptyMap());
+
+        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
+        retryConfiguration.setMaxRetryTime(30);
+        ApiClient.Retry retry = new ApiClient.Retry(retryConfiguration);
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        boolean result = retry.shouldRetry(response);
+        Assert.assertTrue(result, "Even though Retry-After header is missing, it will sleep for 3 Sec by default and returns true");
+        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 3000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 3100);
+        stopwatch.stop();
+    }
+
+    private ApiClient getApiClient(ApiClient.RetryConfiguration retryConfiguration) {
+        region = getEnvironment();
+        if (region == null) {
+            useenum = false;
+        }
+        try {
+            ApiClient.Builder builder = ApiClient.Builder.standard();
+            builder = builder.withProperty(ApiClientConnectorProperty.CONNECTOR_PROVIDER, connector);
+            if (useenum) {
+                builder = builder.withBasePath(region);
+            } else {
+                builder = builder.withBasePath("https://api." + environment);
+            }
+            builder = builder.withRetryConfiguration(retryConfiguration);
+            apiClient = builder.build();
+
+            ApiResponse<AuthResponse> authResponse = apiClient.authorizeClientCredentials(getClientId(), getClientSecret());
```

Comment:
The tests are failing here with an NPE.

```
[INFO] Running TestSuite
[TEST] beforeTest
Before test
[SUCCESS] beforeTest
[TEST] setup
[SUCCESS] setup
[TEST] invokeTestWith_429
Not in PureCloudRegionHosts using string value
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
java.lang.NullPointerException
[FAILED] invokeTestWith_429
[ERROR] Tests run: 53, Failures: 18, Errors: 0, Skipped: 35, Time elapsed: 0.904 s <<< FAILURE! - in TestSuite
[ERROR] invokeTestWith_429(com.mypurecloud.sdk.v2.ApiClientRetryTest)  Time elapsed: 0.236 s  <<< FAILURE!
java.lang.AssertionError: null
	at com.mypurecloud.sdk.v2.ApiClientRetryTest.getApiClient(ApiClientRetryTest.java:299)
	at com.mypurecloud.sdk.v2.ApiClientRetryTest.invokeTestWith_429(ApiClientRetryTest.java:53)
```

Reply:
I always used AccessToken to test against local because with clientId and clientSecret It would run against https://apps.inindca.com.

But adding this as like in the Integration test(SdkTests).

Reply:
The tests must authenticate using client credentials. The test is configured correctly to do that, but the NPE is coming from your retry code somewhere. use `ex.PrintStackTrace()` to see exactly where. Make sure all of the tests are passing against DCA; none of the tests should rely on the API _actually_ responding with an error; those should be mocked.

Reply:
NPE occurred due to mocked response on API call during authentication that was used for both Authentication and actual API call.

`Mockito.spy()` fixed the issue with a real call for authentication and with a mocked response for Actual API call.